File: OJ practices/LeetCode题目集/82. Remove Duplicates from Sorted List II(medium).java
/*
 * @Author: SourDumplings
 * @Date: 2020-08-27 19:06:10
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode-cn.com/problems/remove-duplicates-from-sorted-list-ii/
 */

class Solution
{
    public ListNode deleteDuplicates(ListNode head)
    {
        ListNode h = new ListNode(0);
        h.next = head;
        ListNode f = h;
        ListNode p = head;
        if (p == null)
        {
            return null;
        }
        ListNode r = head.next;
        if (r == null)
        {
            return head;
        }
        while (true)
        {
            while (r != null && r.val == p.val)
            {
                r = r.next;
            }
            if (r == null)
            {
                if (r != p.next)
                {
                    f.next = null;
                }
                break;
            }
            else if (r != p.next)
            {
                f.next = r;
                p = r;
                r = p.next;
            }
            else
            {
                f = f.next;
                if (f != null)
                {
                    p = f.next;
                }
                else
                {
                    break;
                }
                if (p != null)
                {
                    r = p.next;
                }
                else
                {
                    break;
                }
            }
        }
        return h.next;
    }
}

File: OJ practices/LeetCode题目集/82. Remove Duplicates from Sorted List II(medium).cpp
/**
 * @file 82. Remove Duplicates from Sorted List II(medium).cpp
 * @author SourDumplings (sourdumplings@qq.com)
 * @brief
 * https://leetcode.cn/problems/remove-duplicates-from-sorted-list-ii/description/?envType=study-plan-v2&envId=top-interview-150
 * @version 1.0.0
 * @date 2024-11-28
 *
 * @copyright Copyright (c) 2024 SourDumplings
 *
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution
{
public:
	ListNode* deleteDuplicates(ListNode* head)
	{
		ListNode* tempHead = new ListNode(0, head);
		ListNode* f		   = tempHead;
		ListNode* p		   = head;
		while (f && f->next)
		{
			p = f->next;
			while (p->next && p->next->val == p->val)
			{
				p = p->next;
			}
			if (p != f->next) { f->next = p->next; }
			else { f = f->next; }
		}
		ListNode* ret = tempHead->next;
		delete tempHead;
		return ret;
	}
};
